google/cloud/firestore_v1/pipeline_stages.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2025 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+from typing import Optional
+from abc import ABC
+from abc import abstractmethod
+
+from google.cloud.firestore_v1.types.document import Pipeline as Pipeline_pb
+from google.cloud.firestore_v1.types.document import Value
+from google.cloud.firestore_v1.pipeline_expressions import Expr
+
+
+class Stage(ABC):
+    """Base class for all pipeline stages.
+
+    Each stage represents a specific operation (e.g., filtering, sorting,
+    transforming) within a Firestore pipeline. Subclasses define the specific
+    arguments and behavior for each operation.
+    """
+
+    def __init__(self, custom_name: Optional[str] = None):
+        self.name = custom_name or type(self).__name__.lower()
+
+    def _to_pb(self) -> Pipeline_pb.Stage:
+        return Pipeline_pb.Stage(
+            name=self.name, args=self._pb_args(), options=self._pb_options()
+        )
+
+    @abstractmethod
+    def _pb_args(self) -> list[Value]:
+        """Return Ordered list of arguments the given stage expects"""
+        raise NotImplementedError
+
+    def _pb_options(self) -> dict[str, Value]:
+        """Return optional named arguments that certain functions may support."""
+        return {}
+
+    def __repr__(self):
+        items = ("%s=%r" % (k, v) for k, v in self.__dict__.items() if k != "name")
+        return f"{self.__class__.__name__}({', '.join(items)})"
+
+
+class Collection(Stage):
```

Comment:
FWIW, in most SDKs, we tried to make adding a stage a method on Pipeline. Only in the Java SDKs we expose the Stage classes. I guess it's too early to tell how you are planning to implement the API to add stages to a Pipeline. And I guess you will do whatever is idomatic for Python.

Reply:
Python currently exposes them both ways, but using the methods on pipeline is more idiomatic and user-friendly. Maybe I'll mark this class as private, and keep the pipeline methods as the official public way to add a stage